// @pnp/graph/lists/types.js
import { __decorate } from "tslib";
import { _GraphQueryableCollection, graphInvokableFactory, _GraphQueryableInstance } from "../graphqueryable.js";
import { defaultPath, deleteable, updateable, getById } from "../decorators.js";
import { graphPost } from "../operations.js";
import { body } from "@pnp/queryable";
/**
 * Represents a booking service entity
 */
let _List = class _List extends _GraphQueryableInstance {
};
_List = __decorate([
    deleteable(),
    updateable()
], _List);
export { _List };
export const List = graphInvokableFactory(_List);
/**
 * Describes a collection of booking service objects
 *
 */
let _Lists = class _Lists extends _GraphQueryableCollection {
    /**
     * Create a new booking service as specified in the request body.
     *
     * @param list  a JSON representation of a List object.
     */
    async add(list) {
        const data = await graphPost(this, body(list));
        return {
            data,
            list: this.getById(data.id),
        };
    }
};
_Lists = __decorate([
    defaultPath("services"),
    getById(List)
], _Lists);
export { _Lists };
export const Lists = graphInvokableFactory(_Lists);
