// @pnp/graph/sites/types.js
import { __decorate } from "tslib";
import { defaultPath } from "../decorators.js";
import { graphInvokableFactory, _GraphQueryableCollection, _GraphQueryableInstance } from "../graphqueryable.js";
/**
 * Sites
 */
let _Sites = class _Sites extends _GraphQueryableCollection {
    /**
     * Gets the team site for the group
     */
    get root() {
        return Site(this, "root");
    }
    getById(id) {
        return Site(this, id);
    }
    /**
     * Get a Site by URL
     * @param hostname: string, the host of the site e.g. "contoso.sharepoint.com"
     * @param siteUrl: string, the server relative url of the site e.g. "/sites/teamsite1"
     * @returns ISite
    */
    getByUrl(hostname, siteUrl) {
        return Site(this, `${hostname}:${siteUrl}:`);
    }
};
_Sites = __decorate([
    defaultPath("sites")
], _Sites);
export { _Sites };
export const Sites = graphInvokableFactory(_Sites);
/**
 * Site
 */
export class _Site extends _GraphQueryableInstance {
    get sites() {
        return Sites(this);
    }
}
export const Site = graphInvokableFactory(_Site);
