// @pnp/graph/onedrive/users.js
import { addProp, body } from "@pnp/queryable";
import { graphPost } from "../operations.js";
import { _User } from "../users/types.js";
import { Drive, Drives, _Drive, DriveItem, _DriveItem } from "./types.js";
addProp(_User, "drive", Drive);
addProp(_User, "drives", Drives);
/**
 * Get special folder (as drive) for a user.
 */
_Drive.prototype.special = function special(specialFolder) {
    return DriveItem(this, `special/${specialFolder}`);
};
export var SpecialFolder;
(function (SpecialFolder) {
    SpecialFolder["Documents"] = "documents";
    SpecialFolder["Photos"] = "photos";
    SpecialFolder["CameraRoll"] = "cameraroll";
    SpecialFolder["AppRoot"] = "approot";
    SpecialFolder["Music"] = "music";
})(SpecialFolder || (SpecialFolder = {}));
_DriveItem.prototype.restore = function restore(restoreOptions) {
    return graphPost(DriveItem(this, "restore"), body(restoreOptions));
};
_DriveItem.prototype.follow = function follow() {
    return graphPost(DriveItem(this, "follow"), null);
};
_DriveItem.prototype.unfollow = function unfollow() {
    return graphPost(DriveItem(this, "unfollow"), null);
};
