// @pnp/graph/outlook/types.js
import { __decorate } from "tslib";
import { _GraphQueryableCollection, _GraphQueryableInstance, graphInvokableFactory } from "../graphqueryable.js";
import { defaultPath, deleteable, getById, updateable } from "../decorators.js";
import { graphPost } from "@pnp/graph";
import { body } from "@pnp/queryable";
/**
 * Outlook
 */
export class _Outlook extends _GraphQueryableInstance {
    get masterCategories() {
        return MasterCategories(this);
    }
}
export const Outlook = graphInvokableFactory(_Outlook);
/**
 * Describes an Outlook Category instance
 */
let _OutlookCategory = class _OutlookCategory extends _GraphQueryableInstance {
};
_OutlookCategory = __decorate([
    deleteable(),
    updateable()
], _OutlookCategory);
export { _OutlookCategory };
export const OutlookCategory = graphInvokableFactory(_OutlookCategory);
/**
 * Categories
 */
let _MasterCategories = class _MasterCategories extends _GraphQueryableCollection {
    /**
     * Adds a new event to the collection
     *
     * @param properties The set of properties used to create the event
     */
    async add(properties) {
        const data = await graphPost(this, body(properties));
        return {
            data,
        };
    }
};
_MasterCategories = __decorate([
    defaultPath("masterCategories"),
    getById(OutlookCategory)
], _MasterCategories);
export { _MasterCategories };
export const MasterCategories = graphInvokableFactory(_MasterCategories);
